src/core/lwip/tcp_in.c

```diff
@@ -241,30 +216,18 @@ void L3_level_tcp_input(struct pbuf *p, struct tcp_pcb *pcb)
                             tcp_abort(pcb);
                             goto aborted;
                         }
-                        pbuf_split_64k(in_data.recv_data, &rest);
                         if (in_data.flags & TCP_PSH) {
                             in_data.recv_data->flags |= PBUF_FLAG_PUSH;
                         }
                         /* Notify application that data has been received. */
                         TCP_EVENT_RECV(pcb, in_data.recv_data, ERR_OK, err);
                         if (err == ERR_ABRT) {
-                            if (rest) {
-                                pbuf_cat(in_data.recv_data, rest); /* undo splitting */
-                            }
                             goto aborted;
                         }
                         /* If the upper layer can't receive this data, store it */
                         if (err != ERR_OK) {
-                            if (rest) {
-                                pbuf_cat(in_data.recv_data, rest); /* undo splitting */
-                            }
-                            pcb->refused_data = in_data.recv_data;
-                            LWIP_DEBUGF(
-                                TCP_INPUT_DEBUG,
-                                ("tcp_input: keep incoming packet, because pcb is \"full\"\n"));
-                            break;
-                        } else {
-                            in_data.recv_data = rest;
+                            pcb->rcv_wnd += in_data.recv_data->tot_len;
```

Comment:
Is there a scenario when upper layer refuses the data? What will happen if it does? will we corrupt TCP stream?

Reply:
the only upper layer with RX support is TLS RX. Even if errors happen and TLS session is terminated, XLIO keeps the buffer and returns ERR_OK. The buffer is probably freed in destructor.

So, most likely scenario to fail the RX is when lwip triggers some error, passes the error to the recv() method and recv() returns it back. This is a confusing scheme.